### ISSUE TITLE
training.seed: a streamed run is bit-reproducible, a resident 4-bit run is not

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#354]
`training.seed` (#341) makes a **streamed** run bit-reproducible and leaves the
**resident** run scattering. Measured on 8xH100, two runs of each config, nothing
changed between them but the run itself.

Llama-3.1-8B, `quantization: 4bit`, task `sft`, `training.seed: 1234`, 64 rows,
1 epoch, batch 2, seq 256, LoRA r=8. The two configs differ **only** in
`stream_layers`.

| config | run 1 `train_loss` | run 2 `train_loss` | reproducible? |
|---|---|---|---|
| `stream_layers: true` | 3.6995859146118164 | 3.6995859146118164 | **yes, to the last digit** |
| `stream_layers: false` | 3.6887452602386475 | 3.686117172241211 | **no** — 0.071% apart |

So the seed is reaching *something* — the streamed path is now perfectly
repeatable, which is a real improvement over the 1.4–1.8% process-to-process
scatter this project recorded before #341 landed. But on the ordinary resident
4-bit path, two runs of one unchanged config with an explicit seed still do not
reproduce each other.

## Why this matters beyond tidiness

Every A/B this project runs through the CLI — quality arms, throughput arms,
`soup ship` verdicts — assumes that a difference between two runs means something.
On the resident path it currently does not, unless the difference clears an
unmeasured noise floor. `benchmarks/gate-h100-validation.md` had to discard two
attempts to demonstrate a real correctness defect end-to-end for exactly this
reason, and the reason was believed to be "the CLI has no seed control". The seed
control now exists and the resident half still scatters, so the diagnosis was
incomplete.

## Not the same as #353

#353 is about `training.seed` reaching only the SFT wrapper and not the other
tasks. This is `task: sft` — the one wrapper that *does* receive it — and the
resident arm still fails to reproduce. Whatever remains unseeded is downstream of
where #353 points.

## Candidate causes, untested

- The `bitsandbytes` NF4 quantisation that happens **at load time** on the resident
  path. The streamed path quantises **offline into shards** and reloads identical
  bytes every run, which would explain exactly this asymmetry: the streamed arm
  cannot vary because its weights were frozen to disk before the run started.
- Anything seeded after the model is built rather than before.

The first is easy to test: hash the resident 4-bit weights on two runs and see
whether they differ before a single step is taken. If they do, the seed is not
the problem and the quantiser is.

## Acceptance

- [ ] Two resident runs of one config with an explicit `training.seed` produce the
      same `train_loss`
- [ ] A test asserting it, for the resident path specifically
- [ ] If load-time quantisation is the cause, it is documented as such — because
      then it also means **a resident 4-bit model is not the fixed reference the
      correctness gates assume it is**

Measured on: torch 2.13.0+cu130, bnb 0.50.0, trl 0.26.2, H100 80GB.
Logs: `/root/logs/bench_e2e_sft{,_resident}_r{1,2}.log`.

**Comment by MakazhanAlpamys** [comment]
## The leading hypothesis in the issue body is wrong — measured, not argued

I proposed load-time NF4 quantisation as the likely cause, and said it was easy to
test. It was, and it is not the cause.

Two resident 4-bit loads of the same checkpoint in one process, hashing the packed
`uint8` weights **before a single training step**:

```
load 1: 224 packed tensors, sha256 5a49a57cbe41a71c10ae47e44bae453e
load 2: 224 packed tensors, sha256 5a49a57cbe41a71c10ae47e44bae453e
```

**Byte-identical.** `bitsandbytes` load-time quantisation is deterministic, so the
resident model *is* a fixed starting point, and the reassuring corollary is that
the correctness gates' resident reference is not compromised — the worry raised in
the issue body is retracted.

The scatter is downstream of the weights.

## What that leaves

Ruled out: the quantiser. Still open, roughly in order of how cheap they are to
test:

1. **Non-deterministic CUDA kernels on the resident path.** This is now the
   leading candidate, and it has a mechanism specific to *this* asymmetry: after
   the #331 repair the streamed path runs `dequantize_4bit` + `F.linear`, while the
   resident path goes through `bitsandbytes.MatMul4Bit`. Those are genuinely
   different kernels, and an atomics-based reduction in one and not the other would
   produce exactly what is measured — one arm reproducible, the other not, same
   seed, same weights. Test: re-run the resident arm under
   `CUBLAS_WORKSPACE_CONFIG=:4096:8` with `torch.use_deterministic_algorithms(True)`
   and see whether the scatter disappears.
2. **LoRA dropout.** The bench config sets no `lora.dropout`, so it takes the
   schema default; if that is non-zero, dropout RNG is in the loop. Cheap to
   eliminate by setting it to 0 explicitly and re-running both arms.
3. Anything seeded after the model is built rather than before.

Note that (1), if confirmed, means the difference is **not a bug in the seed
plumbing at all** — it would mean the resident path is inherently
non-reproducible at the kernel level while the streamed path is not, which is an
unusual thing to be able to say in streaming's favour and would belong in the
record rather than in a fix.

Script: `quant_determinism.py`. Environment as above.

**Comment by MakazhanAlpamys** [comment]
## Diagnosed. `training.seed` cannot reach the adapter init on the resident path — the adapter is already built when the seed is set.

Three candidates, all tested, two eliminated and the third confirmed.

**Eliminated — load-time quantisation** (previous comment): two resident 4-bit
loads hash byte-identically.

**Eliminated — LoRA dropout.** Set `lora.dropout: 0.0` explicitly and re-run the
resident arm: 3.6725127696990967 / 3.6818904876708984. Still scatters, slightly
*more*.

**Eliminated — bitsandbytes kernels.** The cleanest control: resident with
`quantization: none`, no bnb in the path at all. 3.6663429737091064 /
3.650954246520996. **Still scatters**, so it is not a 4-bit phenomenon and the
`MatMul4Bit`-vs-`F.linear` asymmetry I proposed last comment is not it either.

**Confirmed — the adapter is initialised before any seed is set.**

`training.seed` is threaded into `TrainingArguments` and nowhere else
(`trainer/sft.py:418`). `Trainer.__init__` does call `set_seed(args.seed)` — but
`get_peft_model` runs at `trainer/sft.py:833`, i.e. **the LoRA matrices already
exist by the time the Trainer is constructed**. Nothing the seed does afterwards
can change them.

Measured directly — two `get_peft_model` builds in one process, hashing `lora_A`,
no training:

```
build 1  lora_A sha256 b47d4a0b82caa3d4f5cca9d0a57d0312
build 2  lora_A sha256 07846d1f2c6462addf32060759d7c17c
RESULT: DIFFERENT -> LoRA init varies before any seed is set
```

That explains every observation at once: the scatter is independent of
quantisation and of dropout (both downstream of init), and the **streamed path is
reproducible because `build_streamed_model` seeds its own adapter init** — which
this project already documented as a difference between the two paths, but as a
nuisance for A/B comparisons rather than as the reason one half is deterministic
and the other is not.

## What the fix looks like

Call `set_seed(resolved_seed)` **before** the model is built — or at minimum
before `get_peft_model` — rather than leaving it to `Trainer.__init__`. The value
is already resolved at `sft.py:418`, so this is a placement problem, not a
plumbing one.

Worth checking while there: the same ordering applies to every other trainer
wrapper that builds a PEFT model before constructing its Trainer, which is all of
them. That overlaps #353 but is a distinct defect — #353 is about the value not
reaching some tasks, this is about it arriving too late even where it does reach.

## Acceptance (updated)

- [ ] `set_seed` is called before the adapter is created, on every wrapper
- [ ] Two resident runs of one config with an explicit `training.seed` produce the
      same `train_loss`
- [ ] A test asserting **`lora_A` is identical across two builds** at a fixed seed —
      the property, not the loss, since the loss can match for other reasons
- [ ] The streamed path keeps its current reproducibility (it seeds its own init;
      an added global `set_seed` must not change that)

Retracted from this issue: the worry that a resident 4-bit model is not a fixed
reference. It is — the quantiser is deterministic.

Scripts: `quant_determinism.py`, plus the resident arms in
`/root/logs/bench_{nodrop,noquant}_r{1,2}.log`.

**Comment by MakazhanAlpamys** [comment]
## The proposed fix is verified, with the control that makes it a test

Diagnosis said: seed before `get_peft_model` instead of leaving it to
`Trainer.__init__`. Tested on the box, without touching the frozen `src/`:

```
seeded first  : efa72b5f6fcd2c39dd24f1914d10c509
                efa72b5f6fcd2c39dd24f1914d10c509
  -> IDENTICAL

control (none): 9b9305ce27e219c0d1a4f3428b7ac2e2
                8b213b95e04ec344dfe1501c2f469984
  -> DIFFERENT
```

**The control is the point.** Two identical hashes on their own are equally
consistent with "this probe cannot see a difference at all" — which is exactly how
the first run of the v0.72.4 gate produced a green verdict while comparing nothing.
The unseeded arm varying proves the probe has teeth.

So `set_seed(resolved_seed)` before the adapter is created is sufficient. One call,
correctly placed.

## Where it goes

`training.seed` is resolved at `trainer/sft.py:418` for `TrainingArguments`;
`get_peft_model` runs at `trainer/sft.py:833`. The call belongs before the latter —
and the same ordering applies to every wrapper that builds a PEFT model before
constructing its Trainer, which is all of them.

One thing to preserve: the streamed path is **already** reproducible because
`build_streamed_model` seeds its own adapter init. Adding a global `set_seed`
earlier must not disturb that — a test asserting the streamed arm still reproduces
belongs in the same change.

## Summary of the whole investigation

| step | result |
|---|---|
| symptom | streamed reproduces bit-for-bit, resident does not, same seed |
| hypothesis 1 — load-time NF4 quantisation | **eliminated**, byte-identical hashes |
| hypothesis 2 — LoRA dropout | **eliminated**, `dropout: 0.0` still scatters |
| hypothesis 3 — bnb kernel asymmetry | **eliminated**, `quantization: none` still scatters |
| hypothesis 4 — adapter built before seed | **confirmed**, `lora_A` differs across builds |
| proposed fix | **verified**, identical under seed-first, control varies |

Three of my own four hypotheses were wrong. Each is left in the thread with the
control that killed it, because the eliminations are what make the fourth
credible.

**Comment by MakazhanAlpamys** [comment]
## This is not cosmetic: the unseeded adapter init makes `soup ship`'s leg-2 verdicts unreliable

Measured the consequence. Three runs of **one unchanged config** on **one dataset**,
resident 8B, then `soup ship` on each — against five streamed runs of the same
design as a control.

**Task metric (300 held-out items):**

| | streamed (n=5) | resident (n=3) |
|---|---|---|
| values | 0.886667 ×5 | 0.883333 / 0.886667 / 0.890000 |
| **spread** | **0.000000** | **0.006667** |
| distinct adapter hashes | **1** | **3** |

**Leg-2 suites, same runs:**

| suite | resident values | **spread** | streamed spread |
|---|---|---|---|
| `mini_common_sense` | 0.875 / 1.000 / 0.625 | **0.375** | 0.0000 |
| `mini_mmlu` | 0.500 / 0.692 / 0.423 | **0.269** | 0.0000 |
| `mini_tool_call` | 0.150 / 0.075 / 0.150 | 0.075 | 0.0000 |
| `mini_format_json` | 0.875 / 0.875 / 0.925 | 0.050 | 0.0000 |
| `mini_instruction` | 1.000 / 0.958 / 1.000 | 0.042 | 0.0000 |
| `mini_arithmetic` | 1.000 / 1.000 / 0.972 | 0.028 | 0.0000 |
| `mini_safety` | 1.000 / 0.975 / 1.000 | 0.025 | 0.0000 |

### Why this is the important number

`soup ship`'s default `forgetting_threshold` is **0.05**. `mini_common_sense` moves
**0.375** between two runs that differ in nothing a user can see — **7.5x the
threshold**. `mini_mmlu` moves 0.269, **5.4x**. Five of seven suites can cross the
regression threshold on re-running an identical config.

So a leg-2 REGRESSED flag on a resident model is, for those two suites, closer to a
coin flip than to a measurement — and the flag is what turns a SHIP into a
DON'T SHIP.

The streamed column is the control that makes this attributable: identical design,
zero spread everywhere, because `build_streamed_model` seeds its own adapter init.
The entire variance is the unseeded `lora_A`.

### What it changes about the fix

The fix is unchanged — one `set_seed` before `get_peft_model`. What changes is the
priority. This is not "runs should be reproducible for tidiness"; it is "the gate's
leg-2 verdict is not reproducible on the default path", and every A/B this project
runs through the CLI inherits it.

Suggested addition to acceptance:

- [ ] three runs of one config produce the **same leg-2 suite scores**, not merely
      the same `train_loss` — the loss is the least sensitive of the three
      quantities measured here and would pass while the suites still scattered

Caveat: n=3, one model, one dataset. It sizes the effect; it does not calibrate it.
Raw: `/root/results/resnoise_r{1,2,3}.json` against `/root/results/noise_r{1..5}.json`.
